gpgs/src/java/com/defold/gpgs/GpgsJNI.java

```diff
@@ -0,0 +1,444 @@
+                }
+            } else {
+                // Error message
+            }
+        }
+    }
+
+    public void silentLogin() {
+        this.activity.runOnUiThread(new Runnable() {
+            @Override
+            public void run() {
+                GoogleSignInAccount account = GoogleSignIn.getLastSignedInAccount(activity);
+                if (GoogleSignIn.hasPermissions(account, getSignInOptions().getScopeArray())) {
+                    onConnected(account, MSG_SILENT_SIGN_IN);
+                } else {
+                    mGoogleSignInClient.silentSignIn().addOnCompleteListener(activity,
+                        new OnCompleteListener<GoogleSignInAccount>() {
+                            @Override
+                            public void onComplete(@NonNull Task<GoogleSignInAccount> task) {
+                                if (task.isSuccessful()) {
+                                    onConnected(task.getResult(), MSG_SILENT_SIGN_IN);
+                                } else {
+                                    sendSimpleMessage(MSG_SILENT_SIGN_IN,
+                                            "status", STATUS_FAILED,
+                                            "error",
+                                            "Silent sign-in failed");
+                                }
+                            }
+                        });
+                }
+            }
+        });
+    }
+
+    public void login() {
+        Intent intent = mGoogleSignInClient.getSignInIntent();
+        this.activity.startActivityForResult(intent, RC_SIGN_IN);
+    }
+
+    public void logout() {
+        mGoogleSignInClient.signOut().addOnCompleteListener(this.activity,
+            new OnCompleteListener<Void>() {
+                @Override
+                public void onComplete(@NonNull Task<Void> task) {
+                    mSignedInAccount = null;
+                    mPlayer = null;
+                    sendSimpleMessage(MSG_SIGN_OUT, "status", STATUS_SUCCESS, null, null);
+                }
+            });
+    }
+
+    public String getDisplayName() {
+        if (mPlayer == null) {
+            return null;
+        }
+        return mPlayer.getDisplayName();
+    }
+
+    public String getId() {
+        if (mPlayer == null) {
+            return null;
+        }
+        return mPlayer.getPlayerId();
+    }
+
+    public boolean isLoggedIn() {
+        if (mPlayer == null) {
+            return false;
+        }
+        return true;
+    }
+
+    public void setGravityForPopups(int gravity) {
+        mGravity = gravity;
+    }
+
+    //--------------------------------------------------
+    // GoogleDrive (Snapshots)
+
+    // Client used to interact with Google Snapshots.
+    private SnapshotsClient mPlayerSnapshotsClient = null;
+    private Snapshot mPlayerSnapshot = null;
+    private byte[] currentplayerSave = null;
+
+    private int maxCoverImageSize = 819200;
```

Comment:
Where did you get these values from? Is it from the official docs?

Reply:
Yes, I'll add a comment